codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+            streamInfo.childChannel.pipeline().fireUserEventTriggered(new DefaultHttp2ResetFrame(errorCode));
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
+                Http2Headers headers, int streamDependency, short weight, boolean
+                exclusive, int padding, boolean endStream) throws Http2Exception {
+            onHeadersRead(ctx, streamId, headers, padding, endStream);
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                int padding, boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            fireChildReadAndRegister(streamInfo, new DefaultHttp2HeadersFrame(headers, endOfStream, padding));
+        }
+
+        @Override
+        public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            fireChildReadAndRegister(streamInfo, new DefaultHttp2DataFrame(data.retain(), endOfStream, padding));
+            // We return the bytes in bytesConsumed() once the stream channel consumed the bytes.
+            return 0;
+        }
+    }
+
+    static final class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    // This class uses invoker(ctx).invoke* instead of ctx.* to send to the ctx's handler instead
+    // of the 'next' handler.
+    class Http2StreamChannel extends AbstractHttp2StreamChannel {
+        private final Http2Stream stream;
+        boolean onStreamClosedFired;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        protected void doClose() throws Exception {
+            if (!onStreamClosedFired) {
+                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).setStream(this);
+                invoker(ctx).invokeWrite(ctx, resetFrame, ctx.newPromise());
+                invoker(ctx).invokeFlush(ctx);
+            }
+            super.doClose();
+        }
+
+        @Override
+        protected void doWrite(Object msg) {
+            if (!(msg instanceof Http2StreamFrame)) {
+                throw new IllegalArgumentException("Message must be an Http2StreamFrame:" + msg);
```

Comment:
Good catch. Yes, it does look like we need the release. The later throw is doing it properly.